File: python/src/openharness/models/anthropic.py
"""Anthropic Messages API adapter (Claude models)."""

from __future__ import annotations

import json
import os
from collections.abc import AsyncIterator
from typing import Any

from ..errors import ConfigError, ModelError
from ..types import Message, ModelRequest, ModelResponse, ToolCall, Usage
from .base import HTTPClient, Model, ModelEvent, parse_args

ANTHROPIC_BASE_URL = "https://api.anthropic.com/v1"
ANTHROPIC_VERSION = "2023-06-01"
DEFAULT_MAX_TOKENS = 16000


class AnthropicModel(Model):
    provider = "anthropic"

    def __init__(
        self,
        model: str,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        headers: dict[str, str] | None = None,
        timeout: float = 300.0,
        max_retries: int = 3,
    ):
        self.model = model
        key = api_key or os.environ.get("ANTHROPIC_API_KEY")
        if not key:
            raise ConfigError("anthropic: set ANTHROPIC_API_KEY or pass api_key=")
        self.base_url = base_url or os.environ.get("ANTHROPIC_BASE_URL") or ANTHROPIC_BASE_URL
        self.http = HTTPClient(
            self.base_url,
            {
                "x-api-key": key,
                "anthropic-version": ANTHROPIC_VERSION,
                "Content-Type": "application/json",
                **(headers or {}),
            },
            timeout=timeout,
            max_retries=max_retries,
            provider="anthropic",
        )

    async def aclose(self) -> None:
        await self.http.aclose()

    # ------------------------------------------------------------- request

    def build_body(self, req: ModelRequest, stream: bool) -> dict[str, Any]:
        s = req.settings
        body: dict[str, Any] = {
            "model": self.model,
            "max_tokens": s.max_output_tokens or DEFAULT_MAX_TOKENS,
            "messages": self._messages(req.messages),
        }
        system_parts = [req.system] if req.system else []
        system_parts += [m.content for m in req.messages if m.role == "system" and m.content]
        if system_parts:
            body["system"] = "\n\n".join(system_parts)
        if req.tools:
            body["tools"] = [
                {"name": t.name, "description": t.description, "input_schema": t.parameters} for t in req.tools
            ]
            if s.tool_choice:
                tc = {"auto": {"type": "auto"}, "required": {"type": "any"}, "none": {"type": "none"}}.get(
                    s.tool_choice, {"type": "tool", "name": s.tool_choice}
                )
                if s.parallel_tool_calls is False and tc["type"] != "none":
                    tc["disable_parallel_tool_use"] = True
                body["tool_choice"] = tc
            elif s.parallel_tool_calls is False:
                body["tool_choice"] = {"type": "auto", "disable_parallel_tool_use": True}
        if s.temperature is not None:
            body["temperature"] = s.temperature
        if s.top_p is not None:
            body["top_p"] = s.top_p
        if req.output_schema:
            body["output_config"] = {"format": {"type": "json_schema", "schema": req.output_schema.schema}}
        if stream:
            body["stream"] = True
        body.update(s.extra)
        return body

    def _messages(self, messages: list[Message]) -> list[dict[str, Any]]:
        out: list[dict[str, Any]] = []

        def push(role: str, blocks: list[dict[str, Any]]) -> None:
            if not blocks:
                return
            if out and out[-1]["role"] == role:
                out[-1]["content"].extend(blocks)
            else:
                out.append({"role": role, "content": list(blocks)})

        for m in messages:
            if m.role == "system":
                continue
            if m.role == "user":
                push("user", [{"type": "text", "text": m.content or "(empty)"}])
            elif m.role == "tool":
                block: dict[str, Any] = {
                    "type": "tool_result",
                    "tool_use_id": m.tool_call_id,
                    "content": m.content or "(empty)",
                }
                if m.is_error:
                    block["is_error"] = True
                push("user", [block])
            else:
                raw = self.own_raw(m)
                if raw:
                    push("assistant", [dict(b) for b in raw])
                    continue
                blocks: list[dict[str, Any]] = []
                if m.content:
                    blocks.append({"type": "text", "text": m.content})
                for tc in m.tool_calls:
                    blocks.append({"type": "tool_use", "id": tc.id, "name": tc.name, "input": tc.arguments})
                push("assistant", blocks or [{"type": "text", "text": "(empty)"}])
        # The Messages API requires tool_result blocks to come first in a user turn.
        for msg in out:
            if msg["role"] == "user":
                msg["content"].sort(key=lambda b: 0 if b.get("type") == "tool_result" else 1)
        return out

    # ------------------------------------------------------------- response

    def _to_message(self, blocks: list[dict[str, Any]]) -> Message:
        text = "".join(b.get("text", "") for b in blocks if b.get("type") == "text")
        calls = [
            ToolCall(id=b["id"], name=b["name"], arguments=b.get("input") or {})
            for b in blocks
            if b.get("type") == "tool_use"
        ]
        msg = Message.assistant(text, calls)
        msg.raw = {"provider": self.provider, "data": blocks}
        return msg

    @staticmethod
    def _usage(u: dict[str, Any] | None) -> Usage:
        u = u or {}
        inp = (
            (u.get("input_tokens") or 0)
            + (u.get("cache_read_input_tokens") or 0)
            + (u.get("cache_creation_input_tokens") or 0)
        )
        return Usage(input_tokens=inp, output_tokens=u.get("output_tokens") or 0, requests=1)

    async def generate(self, request: ModelRequest) -> ModelResponse:
        data = await self.http.post_json("messages", self.build_body(request, stream=False))
        if data.get("type") == "error":
            raise ModelError(f"anthropic: {data.get('error')}", body=data, provider=self.provider)
        return ModelResponse(
            self._to_message(data.get("content") or []), self._usage(data.get("usage")), data.get("stop_reason")
        )

    async def stream(self, request: ModelRequest) -> AsyncIterator[ModelEvent]:
        blocks: dict[int, dict[str, Any]] = {}
        partial_json: dict[int, str] = {}
        usage = Usage(requests=1)
        stop = None
        async for event, data in self.http.post_sse("messages", self.build_body(request, stream=True)):
            payload = json.loads(data)
            etype = payload.get("type") or event
            if etype == "message_start":
                usage = self._usage(payload.get("message", {}).get("usage"))
            elif etype == "content_block_start":
                idx = payload["index"]
                blocks[idx] = dict(payload["content_block"])
                if blocks[idx].get("type") == "tool_use":
                    partial_json[idx] = ""
            elif etype == "content_block_delta":
                idx, delta = payload["index"], payload["delta"]
                b = blocks.setdefault(idx, {"type": "text", "text": ""})
                dt = delta.get("type")
                if dt == "text_delta":
                    b["text"] = b.get("text", "") + delta["text"]
                    yield ModelEvent("text_delta", {"delta": delta["text"]})
                elif dt == "input_json_delta":
                    partial_json[idx] = partial_json.get(idx, "") + delta.get("partial_json", "")
                elif dt == "thinking_delta":
                    b["thinking"] = b.get("thinking", "") + delta.get("thinking", "")
                elif dt == "signature_delta":
                    b["signature"] = b.get("signature", "") + delta.get("signature", "")
                elif dt == "citations_delta":
                    b.setdefault("citations", []).append(delta.get("citation"))
            elif etype == "content_block_stop":
                idx = payload["index"]
                if idx in partial_json:
                    blocks[idx]["input"] = parse_args(partial_json.pop(idx))
                    b = blocks[idx]
                    yield ModelEvent("tool_call", {"id": b["id"], "name": b["name"], "arguments": b["input"]})
            elif etype == "message_delta":
                stop = (payload.get("delta") or {}).get("stop_reason") or stop
                out_tokens = (payload.get("usage") or {}).get("output_tokens")
                if out_tokens is not None:
                    usage.output_tokens = out_tokens
            elif etype == "error":
                raise ModelError(
                    f"anthropic stream error: {payload.get('error')}", body=payload, provider=self.provider
                )
        ordered = [blocks[i] for i in sorted(blocks)]
        yield ModelEvent("done", response=ModelResponse(self._to_message(ordered), usage, stop))

File: typescript/src/models/anthropic.ts
/** Anthropic Messages API adapter (Claude models). */

import { ConfigError, ModelError } from "../errors.js";
import { Message, Usage, type ModelRequest, type ModelResponse, type ToolCall } from "../types.js";
import { env, HTTPClient, ownRaw, parseArgs, type HTTPOptions, type Model, type ModelEvent } from "./base.js";

export const ANTHROPIC_BASE_URL = "https://api.anthropic.com/v1";
export const ANTHROPIC_VERSION = "2023-06-01";
export const DEFAULT_MAX_TOKENS = 16000;

type Block = Record<string, any>;

export interface AnthropicModelOptions extends HTTPOptions {
  apiKey?: string;
  baseUrl?: string;
  headers?: Record<string, string>;
}

export class AnthropicModel implements Model {
  readonly provider = "anthropic";
  readonly http: HTTPClient;

  constructor(
    readonly model: string,
    options: AnthropicModelOptions = {},
  ) {
    const key = options.apiKey ?? env("ANTHROPIC_API_KEY");
    if (!key) throw new ConfigError("anthropic: set ANTHROPIC_API_KEY or pass apiKey");
    const baseUrl = options.baseUrl ?? env("ANTHROPIC_BASE_URL") ?? ANTHROPIC_BASE_URL;
    this.http = new HTTPClient(
      baseUrl,
      { "x-api-key": key, "anthropic-version": ANTHROPIC_VERSION, "Content-Type": "application/json", ...(options.headers ?? {}) },
      "anthropic",
      options,
    );
  }

  buildBody(req: ModelRequest, stream: boolean): Record<string, unknown> {
    const s = req.settings ?? {};
    const body: Record<string, unknown> = {
      model: this.model,
      max_tokens: s.maxOutputTokens ?? DEFAULT_MAX_TOKENS,
      messages: this.messages(req.messages),
    };
    const system = [req.system, ...req.messages.filter((m) => m.role === "system").map((m) => m.content)].filter(Boolean);
    if (system.length) body.system = system.join("\n\n");
    if (req.tools?.length) {
      body.tools = req.tools.map((t) => ({ name: t.name, description: t.description, input_schema: t.parameters }));
      if (s.toolChoice) {
        const map: Record<string, Block> = { auto: { type: "auto" }, required: { type: "any" }, none: { type: "none" } };
        const tc: Block = map[s.toolChoice] ?? { type: "tool", name: s.toolChoice };
        if (s.parallelToolCalls === false && tc.type !== "none") tc.disable_parallel_tool_use = true;
        body.tool_choice = tc;
      } else if (s.parallelToolCalls === false) {
        body.tool_choice = { type: "auto", disable_parallel_tool_use: true };
      }
    }
    if (s.temperature !== undefined) body.temperature = s.temperature;
    if (s.topP !== undefined) body.top_p = s.topP;
    if (req.outputSchema) body.output_config = { format: { type: "json_schema", schema: req.outputSchema.schema } };
    if (stream) body.stream = true;
    Object.assign(body, s.extra ?? {});
    return body;
  }

  private messages(messages: Message[]): { role: string; content: Block[] }[] {
    const out: { role: string; content: Block[] }[] = [];
    const push = (role: string, blocks: Block[]) => {
      if (!blocks.length) return;
      const last = out[out.length - 1];
      if (last && last.role === role) last.content.push(...blocks);
      else out.push({ role, content: [...blocks] });
    };
    for (const m of messages) {
      if (m.role === "system") continue;
      if (m.role === "user") push("user", [{ type: "text", text: m.content || "(empty)" }]);
      else if (m.role === "tool") {
        const block: Block = { type: "tool_result", tool_use_id: m.toolCallId, content: m.content || "(empty)" };
        if (m.isError) block.is_error = true;
        push("user", [block]);
      } else {
        const raw = ownRaw(this.provider, m) as Block[] | undefined;
        if (raw) {
          push("assistant", raw.map((b) => ({ ...b })));
          continue;
        }
        const blocks: Block[] = [];
        if (m.content) blocks.push({ type: "text", text: m.content });
        for (const tc of m.toolCalls ?? []) blocks.push({ type: "tool_use", id: tc.id, name: tc.name, input: tc.arguments });
        push("assistant", blocks.length ? blocks : [{ type: "text", text: "(empty)" }]);
      }
    }
    // The Messages API requires tool_result blocks to come first in a user turn.
    for (const msg of out) {
      if (msg.role === "user") msg.content.sort((a, b) => (a.type === "tool_result" ? 0 : 1) - (b.type === "tool_result" ? 0 : 1));
    }
    return out;
  }

  private toMessage(blocks: Block[]): Message {
    const text = blocks.filter((b) => b.type === "text").map((b) => b.text ?? "").join("");
    const calls: ToolCall[] = blocks.filter((b) => b.type === "tool_use").map((b) => ({ id: b.id, name: b.name, arguments: b.input ?? {} }));
    const msg = Message.assistant(text, calls);
    msg.raw = { provider: this.provider, data: blocks };
    return msg;
  }

  private usage(u: any): Usage {
    const input = (u?.input_tokens ?? 0) + (u?.cache_read_input_tokens ?? 0) + (u?.cache_creation_input_tokens ?? 0);
    return new Usage({ inputTokens: input, outputTokens: u?.output_tokens ?? 0, requests: 1 });
  }

  async generate(req: ModelRequest): Promise<ModelResponse> {
    const data = await this.http.postJSON("messages", this.buildBody(req, false), req.signal);
    if (data.type === "error") throw new ModelError(`anthropic: ${JSON.stringify(data.error)}`, undefined, data, this.provider);
    return { message: this.toMessage(data.content ?? []), usage: this.usage(data.usage), stopReason: data.stop_reason };
  }

  async *stream(req: ModelRequest): AsyncIterable<ModelEvent> {
    const blocks = new Map<number, Block>();
    const partialJson = new Map<number, string>();
    let usage = new Usage({ requests: 1 });
    let stop: string | undefined;
    for await (const { event, data } of this.http.postSSE("messages", this.buildBody(req, true), req.signal)) {
      const p = JSON.parse(data);
      const type = p.type ?? event;
      if (type === "message_start") usage = this.usage(p.message?.usage);
      else if (type === "content_block_start") {
        blocks.set(p.index, { ...p.content_block });
        if (p.content_block?.type === "tool_use") partialJson.set(p.index, "");
      } else if (type === "content_block_delta") {
        const b = blocks.get(p.index) ?? { type: "text", text: "" };
        blocks.set(p.index, b);
        const d = p.delta ?? {};
        if (d.type === "text_delta") {
          b.text = (b.text ?? "") + d.text;
          yield { type: "text_delta", delta: d.text };
        } else if (d.type === "input_json_delta") partialJson.set(p.index, (partialJson.get(p.index) ?? "") + (d.partial_json ?? ""));
        else if (d.type === "thinking_delta") b.thinking = (b.thinking ?? "") + (d.thinking ?? "");
        else if (d.type === "signature_delta") b.signature = (b.signature ?? "") + (d.signature ?? "");
        else if (d.type === "citations_delta") (b.citations ??= []).push(d.citation);
      } else if (type === "content_block_stop") {
        if (partialJson.has(p.index)) {
          const b = blocks.get(p.index)!;
          b.input = parseArgs(partialJson.get(p.index));
          partialJson.delete(p.index);
          yield { type: "tool_call", id: b.id, name: b.name, arguments: b.input };
        }
      } else if (type === "message_delta") {
        stop = p.delta?.stop_reason ?? stop;
        if (p.usage?.output_tokens !== undefined) usage.outputTokens = p.usage.output_tokens;
      } else if (type === "error") {
        throw new ModelError(`anthropic stream error: ${JSON.stringify(p.error)}`, undefined, p, this.provider);
      }
    }
    const ordered = [...blocks.entries()].sort(([a], [b]) => a - b).map(([, b]) => b);
    yield { type: "done", response: { message: this.toMessage(ordered), usage, stopReason: stop } };
  }
}
